File: katas/python/SOLID/isp/isp_good.py
# Cumplimiento del ISP: Dividir interfaz ancha en interfaces más pequeñas y enfocadas
# Solución: Crear interfaces separadas para diferentes capacidades

from abc import ABC, abstractmethod


# Interfaces separadas para diferentes capacidades ✅
class Workable(ABC):
    @abstractmethod
    def work(self) -> str:
        pass


class Eatable(ABC):
    @abstractmethod
    def eat(self) -> str:
        pass


class Sleepable(ABC):
    @abstractmethod
    def sleep(self) -> str:
        pass


# Human implementa todas las interfaces (necesita todas las capacidades) ✅
class Human(Workable, Eatable, Sleepable):
    def work(self) -> str:
        return "Humano trabajando"

    def eat(self) -> str:
        return "Humano comiendo"

    def sleep(self) -> str:
        return "Humano durmiendo"


# Robot solo implementa lo que necesita ✅
class Robot(Workable):
    def work(self) -> str:
        return "Robot trabajando"
    # ✅ ¡No necesita implementar eat() o sleep()!


# SuperHuman puede trabajar y tiene habilidades especiales
class SuperHuman(Workable, Sleepable):
    def work(self) -> str:
        return "SuperHumano trabajando a super velocidad"

    def sleep(self) -> str:
        return "SuperHumano durmiendo brevemente"
    # ✅ No necesita comer (obtiene energía del sol)


# Uso - ¡sin más implementaciones forzadas!
if __name__ == "__main__":
    human = Human()
    robot = Robot()
    super_human = SuperHuman()

    print(human.work())  # ✅ Funciona
    print(human.eat())  # ✅ Funciona
    print(human.sleep())  # ✅ Funciona

    print(robot.work())  # ✅ Funciona
    # robot.eat() - El método no existe (¡seguro en tiempo de compilación!)

    print(super_human.work())  # ✅ Funciona
    print(super_human.sleep())  # ✅ Funciona

File: katas/typescript/src/SOLID/isp/isp-good.ts
// Cumplimiento del ISP: Dividir interfaz ancha en interfaces más pequeñas y enfocadas
// Solución: Crear interfaces separadas para diferentes capacidades

// Interfaces separadas para diferentes capacidades ✅
interface Workable {
  work(): string;
}

interface Eatable {
  eat(): string;
}

interface Sleepable {
  sleep(): string;
}

// Human implementa todas las interfaces (necesita todas las capacidades) ✅
class Human implements Workable, Eatable, Sleepable {
  public work(): string {
    return "Humano trabajando";
  }

  public eat(): string {
    return "Humano comiendo";
  }

  public sleep(): string {
    return "Humano durmiendo";
  }
}

// Robot solo implementa lo que necesita ✅
class Robot implements Workable {
  public work(): string {
    return "Robot trabajando";
  }
  // ✅ ¡No necesita implementar eat() o sleep()!
}

// SuperHuman puede trabajar y tiene habilidades especiales
class SuperHuman implements Workable, Sleepable {
  public work(): string {
    return "SuperHumano trabajando a super velocidad";
  }

  public sleep(): string {
    return "SuperHumano durmiendo brevemente";
  }
  // ✅ No necesita comer (obtiene energía del sol)
}

// Uso - ¡sin más implementaciones forzadas!
const human = new Human();
const robot = new Robot();
const superHuman = new SuperHuman();

console.log(human.work()); // ✅ Funciona
console.log(human.eat()); // ✅ Funciona
console.log(human.sleep()); // ✅ Funciona

console.log(robot.work()); // ✅ Funciona
// robot.eat() - El método no existe (¡seguro en tiempo de compilación!)

console.log(superHuman.work()); // ✅ Funciona
console.log(superHuman.sleep()); // ✅ Funciona

export { Workable, Eatable, Sleepable, Human, Robot, SuperHuman };
